src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,612 @@
+/-
+Copyright (c) 2021 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import measure_theory.vector_measure
+import order.symm_diff
+
+/-!
+# Hahn decomposition
+
+This file defines positive and negative sets with respect to a vector measure and
+prove the Hahn decomposition theorem (signed version).
+
+## Main definitions
+
+* `vector_measure.positive` : a set `i` is positive with respect to a vector measure `v` if `v`
+  restricted on `i` is non-negative, i.e. `0 ≤ v.restrict i`.
+* `vector_measure.negative` : a set `i` is negative with respect to a vector measure `v` if `v`
+  restricted on `i` is non-positive, i.e. `v.restrict i ≤ 0`.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition theorem
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+variables {α β : Type*} [measurable_space α]
+variables {M : Type*} [add_comm_monoid M] [topological_space M] [ordered_add_comm_monoid M]
+
+namespace measure_theory
+
+namespace vector_measure
+
+/-- A set `i` is positive with respect to a vector measure `v` if `v` restricted
+on`i` is non-negative, i.e. `0 ≤ v.restrict i`. -/
+def positive (v : vector_measure α M) (i : set α) : Prop :=
+0 ≤ v.restrict i
```

Comment:
Does it make things need fewer `convert`s if this definition is as follows?
```suggestion
restrict 0 i ≤ v.restrict i
```

Reply:
right, we won't need to convert then. In that case, should we even have the lemmas that can be generalized since you may just apply those lemmas directly?

Reply:
That's what I was leaning towards :). Maybe it's worth having shorhand notation for `w,restrict i ≤ v.restrict i`, something like `w ≤[i] v`, so that you don't even need `positive` and `negative`, and just write them `0 ≤[i] v` and `v ≤[i] 0`.

I think there's a notation something like this elsewhere in the measure theory library?

Reply:
Do you think the notation `w ≤[i] v` should be in a locale?

Reply:
also, one problem with this notation is that we will have to write `(0 : vector_measure α M) ≤[i] v` since lean can't infer 0 is a vector measure. What do you think about keeping the definitions but removing the lemmas? I think keeping the definitions is a good idea since its shorter and easier to search for

Reply:
https://github.com/leanprover-community/mathlib/pull/8388#discussion_r681682817